### PR TITLE
Bug/response validation

### PR DESCRIPTION
**Pull request by mikerott**
See description from prior attempt:  https://github.com/asecurityteam/nexpose-asset-attributor/pull/18

This PR solves it in a better way in that now the GET responses will always include all fields by having the custom unmarshalling from the DB contents ensure all struct fields are non-nil, with a test to ensure it.  This is desirable over omitting empty fields in the response because the end-consumer (a human) will want to see all fields to simplify the GET/cut/correct/paste/POST workflow.

### CODE REVIEW

File: pkg/domain/attributor_test.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+	fmt.Println(string(marshalled))
+
+	fmt.Println("EXPECTED")
+	fmt.Println(expected)
+	marshalled2, _ := json.Marshal(expected)
+	fmt.Println(string(marshalled2))
+
+	require.True(t, reflect.DeepEqual(expected, partial), "marshalled object does not equal expected object")
+}
```
Comment (faldridge) [inline pkg/domain/attributor_test.go]: Don't we need to test the non-empty case as well?

Comment (mikerott) [reply pkg/domain/attributor_test.go]: yeah.  That was me being lazy.  I'll write up that test now.

File: pkg/domain/attributor.go [hunk 2]
```diff
@@ -42,6 +44,57 @@ type NexposeAttributedAssetVulnerabilities struct {
...
 }
 
+// UnmarshalJSON is a custom unmarshaller to ensure no `nil` or `null` fields in the resulting struct
+func (n *NexposeAttributedAssetVulnerabilities) UnmarshalJSON(data []byte) error {
+	type Alias NexposeAttributedAssetVulnerabilities
+	aux := &struct {
+		*Alias
+	}{
+		Alias: (*Alias)(n),
+	}
```
Comment (LorenBoren) [inline pkg/domain/attributor.go]: Q: Wait, what just happened?

Comment (faldridge) [reply pkg/domain/attributor.go]: Yes, please.

File: pkg/domain/attributor_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			Tags:               make(map[string]string),
+		},
+	}
+
+	// kind of a dumb test... but good enough to see
+	marshalled, _ := json.Marshal(partial)
+	badMarshalMsg := "You've added a new field in the struct hierarchy that is type array, slice, or map, but forgot to add custom unmarshalling logic for it"
+	require.False(t, strings.Contains(string(marshalled), "null"), badMarshalMsg)
+
+	fmt.Println(string(marshalled))
```
Comment (faldridge) [inline pkg/domain/attributor_test.go]: Task: remove the `fmt.Println` statements.

File: pkg/domain/attributor_test.go (hunk 1 above)
Comment (mikerott) [reply pkg/domain/attributor_test.go]: Done

File: pkg/domain/attributor_test.go (hunk 3 above)
Comment (mikerott) [reply pkg/domain/attributor_test.go]: Done

File: pkg/domain/attributor.go (hunk 2 above)
Comment (mikerott) [reply pkg/domain/attributor.go]: I have to use a type alias because I'm actually calling `json.Unmarshal` _inside_ the `UnmarshalJSON` function (see a few lines down).  That risks an infinite recursion because the target memory address in that `json.Unmarshal` function call _is the struct to which this custom unmarshaller function is attached_.

The `Alias` therefore is a very temporary type that does _not_ have the custom `UnmarshalJSON` function attached (because type aliases do not inherit attached functions), thus avoiding the infinite recursion.
